### PR TITLE
feat: Allow for zero-width fixed size lists

### LINKED ISSUE #18921: Zero-sized Array caused panic exception
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
pl.Series('foo', [None], pl.Array(pl.Null, 0))
```

### Log output

_No response_

### Issue description

Attempting to create a zero-width array causes a panic

### Expected behavior

No panic

### Installed versions

<details>

```
Replace this line with the output of pl.show_versions(). Leave the backticks in place.
```

</details>

### PR DESCRIPTION
**Pull request by coastalwhite**
This PR properly implements `polars.Array(..., 0)`. This does make a few
fundamental changes to how `FixedSizeListArray` works internally. Mainly, it
now keeps a `length` field that actually determines the length instead of
relying on the `values.len() / size`.

Fixes #18921, #16878.

### CODE REVIEW

File: crates/polars-core/src/series/from.rs [hunk 1]
```diff
@@ -517,37 +517,33 @@ unsafe fn to_physical_and_dtype(
             to_physical_and_dtype(out, md)
         },
         #[cfg(feature = "dtype-array")]
-        #[allow(unused_variables)]
         ArrowDataType::FixedSizeList(field, size) => {
-            feature_gated!("dtype-array", {
```
Comment (coastalwhite) [inline crates/polars-core/src/series/from.rs:522]: drive-by: This is already feature-gated, no real sense in feature gating it again

File: py-polars/tests/unit/io/test_parquet.py [hunk 2]
```diff
@@ -924,7 +923,7 @@ def test_parquet_array_dtype_nulls() -> None:
                 [[]],
                 [[None]],
                 [[[None], None]],
-                [[[None], []]],
+                [[[None], [None]]],
```
Comment (coastalwhite) [inline py-polars/tests/unit/io/test_parquet.py:926]: This was a clear error

File: py-polars/tests/unit/operations/test_reshape.py [hunk 3]
```diff
@@ -100,24 +100,14 @@ def test_reshape_empty_valid_1d(shape: tuple[int, ...]) -> None:
...
-def test_reshape_empty_invalid_2d(shape: tuple[int, ...]) -> None:
-    s = pl.Series("a", [], dtype=pl.Int64)
-    with pytest.raises(
-        InvalidOperationError,
-        match=re.escape(
-            f"cannot reshape empty array into shape {display_shape(shape)}"
-        ),
-    ):
-        s.reshape(shape)
-
```
Comment (coastalwhite) [inline py-polars/tests/unit/operations/test_reshape.py:113]: None of the cases now fail

File: crates/polars-arrow/src/array/static_array.rs [hunk 4]
```diff
@@ -392,13 +391,6 @@ impl StaticArray for FixedSizeListArray {
     fn full_null(length: usize, dtype: ArrowDataType) -> Self {
         Self::new_null(dtype, length)
     }
-
-    fn full(length: usize, value: Self::ValueT<'_>, dtype: ArrowDataType) -> Self {
-        let singular_arr = FixedSizeListArray::new(dtype, value, None);
-        let mut arr = GrowableFixedSizeList::new(vec![&singular_arr], false, length);
-        unsafe { arr.extend_copies(0, 0, 1, length) }
-        arr.into()
-    }
```
[the remark this answers is not in the document]
Comment (orlp) [reply crates/polars-arrow/src/array/static_array.rs:401]: If I remember correctly this avoids going through `Box<dyn Array>` a bunch.

[the remark this answers is not in the document]
Comment (coastalwhite) [reply crates/polars-arrow/src/array/static_array.rs:401]: Ah okay, I was trying to get this to work and I could not really get it to work. I will revert this.

File: crates/polars-arrow/src/compute/cast/mod.rs [hunk 5]
```diff
@@ -174,25 +174,20 @@ fn cast_list_to_fixed_size_list<O: Offset>(
...
+        for (i, offset) in offsets.iter().enumerate() {
+            is_valid &= offset.to_usize() == start_offset + i * size;
         }
+
+        polars_ensure!(is_valid, ComputeError: "not all elements have the specified width {size}");
+
+        let sliced_values = list
+            .values()
+            .sliced(start_offset, list.offsets().range().to_usize());
+        cast(sliced_values.as_ref(), inner.dtype(), options)?
```
[the remark this answers is not in the document]
Comment (orlp) [reply crates/polars-arrow/src/compute/cast/mod.rs:190]:
```rust
let is_valid = offsets.iter().enumerate().all(|(i, offset)| offset.to_usize() == start_offset + i * size);
```

**Review by ritchie46 — APPROVED** [review]

**Comment by etiennebacher** [comment]
Hi, just FYI you need to repeat the "Fixes" for each issue number to automatically close multiple issues. Currently #16878 is not closed ;)
